sqlx-core/src/postgres/types/ranges.rs

```diff
@@ -0,0 +1,290 @@
+    decode::{TypedDecode, Decode},
+    encode::{Encode, IsNull},
+    postgres::{types::PgRange, PgArgumentBuffer, PgTypeInfo, PgValueFormat, PgValueRef, Postgres},
+};
+use byteorder::{NetworkEndian, ReadBytesExt, WriteBytesExt};
+use core::{
+    convert::TryInto,
+    ops::{Bound, Range, RangeFrom, RangeInclusive, RangeTo, RangeToInclusive},
+};
+
+bitflags::bitflags! {
+  struct RangeFlags: u8 {
+      const EMPTY = 0x01;
+      const LB_INC = 0x02;
+      const UB_INC = 0x04;
+      const LB_INF = 0x08;
+      const UB_INF = 0x10;
+      const LB_NULL = 0x20;
+      const UB_NULL = 0x40;
+      const CONTAIN_EMPTY = 0x80;
+  }
+}
+
+
+impl<'a, T> Decode<'a, Postgres> for PgRange<T>
+where
+    T: for<'b> TypedDecode<'b, Postgres> + 'a,
+{
+    fn accepts(ty: &PgTypeInfo) -> bool {
+        [
+            PgTypeInfo::INT4_RANGE,
+            PgTypeInfo::NUM_RANGE,
+            PgTypeInfo::TS_RANGE,
+            PgTypeInfo::TSTZ_RANGE,
+            PgTypeInfo::DATE_RANGE,
+            PgTypeInfo::INT8_RANGE,
+        ]
+        .contains(ty)
+    }
+
+    fn decode(value: PgValueRef<'a>) -> Result<PgRange<T>, crate::error::BoxDynError> {
+        match value.format() {
+            PgValueFormat::Binary => {
+                decode_binary(value.as_bytes()?, value.format, value.type_info)
+            }
+            PgValueFormat::Text => {
+                decode_str(value.as_str()?, value.format(), value.type_info)
+            }
+        }
+    }
+}
+
+impl<'a, T> Encode<'a, Postgres> for PgRange<T>
+where
+    T: for<'b> Encode<'b, Postgres> + 'a,
+{
+    fn encode_by_ref(&self, buf: &mut PgArgumentBuffer) -> IsNull {
+        let mut flags = match self.start {
+            Bound::Included(_) => RangeFlags::LB_INC,
+            Bound::Excluded(_) => RangeFlags::empty(),
+            Bound::Unbounded => RangeFlags::LB_INF,
+        };
+
+        flags |= match self.end {
+            Bound::Included(_) => RangeFlags::UB_INC,
+            Bound::Excluded(_) => RangeFlags::empty(),
+            Bound::Unbounded => RangeFlags::UB_INF,
+        };
+
+        buf.write_u8(flags.bits()).unwrap();
+
+        let mut write = |bound: &Bound<T>| -> IsNull {
+            match bound {
+                Bound::Included(ref value) | Bound::Excluded(ref value) => {
+                    buf.write_u32::<NetworkEndian>(0).unwrap();
+                    let prev = buf.len();
+                    if let IsNull::Yes = Encode::<Postgres>::encode(value, buf) {
+                        return IsNull::Yes;
+                    }
+                    let len = buf.len() - prev;
+                    buf[prev - 4..prev].copy_from_slice(&(len as u32).to_be_bytes());
+                }
+                Bound::Unbounded => {}
+            }
+            IsNull::No
+        };
+
+        if let IsNull::Yes = write(&self.start) {
+            return IsNull::Yes;
+        }
+        write(&self.end)
+    }
+
+    fn produces(&self) -> PgTypeInfo {
+        todo!()
```

Comment:
The `produces` method still expects a valid Type for `PgRange<T>`.
It is possible to change `fn produces(&self) -> PgTypeInfo` to `fn produces(&self) -> Option<PgTypeInfo>` or `fn produces(&self) -> Result<PgTypeInfo>`.

Reply:
`Encode::produces` is important as we have to communicate the type ID to PostgreSQL. 

Additionally, I'd like to keep the ability to make that type ID deterministic on the value and not just the type.

You don't really have a great value you could put there. 

How about this, let's make this a provided method on the `Encode` trait, have it return `Option<DB::TypeInfo>` and default it to `None` on the trait. 

Within `Arguments` we would go `value.produces().unwrap_or_else(T::type_info)` where currently its just `value.produces()`.

This lets someone override the type-based ID with a value-determined ID if they want but the vast majority of the time where a type-based ID makes sense, it can continue to do so.

Reply:
Ok, thanks! This approach works greatly and will be the one that will be pursued. For future references, I would like to keep this information:

The separation of concerns already described would keep each trait within its responsibility, preventing a lot of runtime checks of the optional `Option<DB::TypeInfo>`.

```rust
pub trait Encode<'q, DB: Database>: Sized {
  // Other methods
}

pub trait TypedEncode<'q, DB: Database>: Encode<'q, DB> + Sized + Type<DB> {
    fn produces(&self) -> DB::TypeInfo {
        Self::type_info()
    }
}
```